presto-geospatial/src/main/java/com/facebook/presto/geospatial/GeoFunctions.java

```diff
@@ -0,0 +1,532 @@
+    @SqlType(StandardTypes.DOUBLE)
+    public static double stY(@SqlType(GEOMETRY) Slice input)
+    {
+        OGCGeometry geometry = deserialize(input);
+        if (geometry.geometryType().equals(POINT)) {
+            return ((OGCPoint) geometry).Y();
+        }
+        throw new IllegalArgumentException("st_y only applies to point. Input type is: " + geometry.geometryType());
+    }
+
+    @Description("Returns the boundary geometry of input geometry")
+    @ScalarFunction
+    @SqlType(GEOMETRY)
+    public static Slice stBoundary(@SqlType(GEOMETRY) Slice input)
+    {
+        OGCGeometry geometry = deserialize(input);
+        OGCGeometry bound = geometry.boundary();
+        if (bound.geometryType().equals(MULTILINE) && ((OGCMultiLineString) bound).numGeometries() == 1) {
+            bound = ((OGCMultiLineString) bound).geometryN(0);
+        }
+        return serialize(bound);
+    }
+
+    @Description("Returns envelope of the input geometry")
+    @ScalarFunction
+    @SqlType(GEOMETRY)
+    public static Slice stEnvelope(@SqlType(GEOMETRY) Slice input)
+    {
+        OGCGeometry geometry = deserialize(input);
+        SpatialReference reference = geometry.getEsriSpatialReference();
+        Envelope envelope = new Envelope();
+        geometry.getEsriGeometry().queryEnvelope(envelope);
+        return serialize(OGCGeometry.createFromEsriGeometry(envelope, reference));
+    }
+
+    @Description("Returns geometry difference of left geometry and right geometry")
+    @ScalarFunction
+    @SqlType(GEOMETRY)
+    public static Slice stDifference(@SqlType(GEOMETRY) Slice left, @SqlType(GEOMETRY) Slice right)
+    {
+        OGCGeometry leftGeometry = deserialize(left);
+        OGCGeometry rightGeometry = deserialize(right);
+        return serialize(leftGeometry.difference(rightGeometry));
+    }
+
+    @Description("Returns distance between left geometry and right geometry")
+    @ScalarFunction("st_distance")
+    @SqlType(StandardTypes.DOUBLE)
+    public static double stDistance(@SqlType(GEOMETRY) Slice left, @SqlType(GEOMETRY) Slice right)
+    {
+        OGCGeometry leftGeometry = deserialize(left);
+        OGCGeometry rightGeometry = deserialize(right);
+        return leftGeometry.distance(rightGeometry);
+    }
+
+    @Description("Returns the exterior ring of the polygon")
+    @ScalarFunction
+    @SqlType(GEOMETRY)
+    public static Slice stExteriorRing(@SqlType(GEOMETRY) Slice input)
+    {
+        OGCGeometry geometry = deserialize(input);
+        if (geometry.geometryType().equals(POLYGON)) {
+            try {
+                Method method = OGCPolygon.class.getMethod("exteriorRing");
+                OGCLineString exteriorRing = (OGCLineString) method.invoke(geometry);
+                return serialize(exteriorRing);
+            }
+            catch (Exception e) {
+                throw new IllegalArgumentException("st_exterior_ring only applies to polygon. Input type is: " + geometry.geometryType());
```

Comment:
Given that the exception occurred after the check for geometry type has succeeded, the error message doesn't make sense.